src/data_race.rs

```diff
@@ -597,7 +597,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         // of the time, based on `rate`.
         let rate = this.machine.cmpxchg_weak_failure_rate;
         let cmpxchg_success = eq.to_scalar()?.to_bool()?
```

Comment:
```suggestion
        let success_rate = 1.0 - this.machine.cmpxchg_weak_failure_rate;
        let cmpxchg_success = eq.to_scalar()?.to_bool()?
            && if can_fail_spuriously {
                this.machine.rng.get_mut().gen::<f64>() < success_rate
            } else {
                true
            };
```
I think this is clearer? We were assigning to a success bool using a failure rate so we should invert the rate first, also implications should be written out as `if` expressions

Reply:
I agree about the implication; no strong opinion about using `>= fail_rate` or `< 1-fail_rate`.

Reply:
Actually we should just use [`gen_bool()`](https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen_bool)

Reply:
Sorry for the delay. I've switched to use gen_bool, and used the suggested change, but can't run tests/checks locally at the moment, so maybe there's some issue.